keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+mechanism. This means that the resource object will stay in existence until all
+garbage collection mechanisms (owner reference, usage references, TTL) agree it
+should be deleted.
+
+[detailed design of TTL]: https://goo.gl/9AUQ3n
+[detailed design of GC after use (part 1)]: https://goo.gl/ws9t3b
+[detailed design of GC after use (part 2)]: https://goo.gl/KeHPiQ
+
+### Generic TTL-based Garbage Collection
+
+First, we need to clarify what TTL means in this context. TTL represents the
+earliest safe time when a resource can be garbage collected. 
+
+Generic TTL-based garbage collection will allow the user to set a TTL on each of
+Kubernetes resources. The TTL starts when a resource is eligible for garbage
+collection and stops when the resource is no longer eligible; the system will
+only garbage collect the resource after the TTL expires. This is useful when the
+user expects a resource to be needed intermittently and/or want to keep the
+resource around for a while before garbage collection kicks in.
+
+Note that we will not use etcd TTL to support this feature, because etcd TTLs
+won’t respect other lifecycle guarantees (e.g. finalizers). This also allows the
+feature to be independent of etcd.
+
+### Clean Up Resources After Use
+
+A resource is no longer *in use* either when it finishes execution (such as a
+terminated Pod or a completed Job), or when it’s not referenced/used by other
+resources (such as a Secret referenced in a PodSpec of any active Pods or
+workload API resources like Deployments).
+
+The garbage collector is responsible for checking whether a resource is still in
+use or not. To do this, the garbage collector will count "usage reference" of a
+resource object, instead of using owner references directly. Not using owner
+references here provides several benefits:
+
+* It’s quick and easy to enable and then disable this feature without updating
+  owner references lists in resources. 
+* Does not require frequent updates, because a resource object is only updated
+  with it’s eligible to be garbage collected.
+* Avoid the problem of unbounded list of owner references.
+
+When a resource is no longer in use, and when the resource’s owners
+are all gone, the garbage collector sets a timestamp on the resource. The
+garbage collector will later clean up those resources based on their timestamps.
+This is similar to [mark-and-sweep][] garbage collection. 
+
+This timestamp represents the last time this resource is needed and shouldn’t be
+garbage collected. It will be empty when the resource is still in use, and will
+only be set when this resource is no longer needed. This is important, so that
+we can avoid the need for frequent updates (like a heartbeat) which could cause
+performance and scalability issues. The timestamp only needs to be updated when
+a resource is either in use or not in use.
+
+Combining with TTL, users can specify the earliest safe time to clean up
```

Comment:
Adding default ttl should be fine, as it'll only make GC take longer before deleting things. 

I think what you meant is making resources being cleaned up by default (`metadata.gcifunused = true`) with some ttl value? I'm worried that it would break people unexpectedly and there's no way to recover from delete. 

In the "Future Work - Namespace-wide Garbage Collection Policy" section in this KEP, I proposed allowing users to create a "clean up policy" object that will set a default cleanup policy for resources. We can also create a global policy resource or allowing cluster admins to use a flag to set a global default value.

Reply:
If you set default GC delay by gcttl will it affect cascading deletion?